internal/controllers/sacd_helpers.go

```diff
@@ -0,0 +1,146 @@
+	if (globalGrants != nil && globalGrants.Contains(tokenclaims.GlobalIdentifier)) || (sourceGrants != nil && sourceGrants.Contains(tokenclaims.GlobalIdentifier)) {
+		return nil
+	}
+	var missingIDs []string
+	for _, reqID := range requestedIDs {
+		if (globalGrants != nil && !globalGrants.Contains(reqID)) && (sourceGrants != nil && !sourceGrants.Contains(reqID)) {
+			missingIDs = append(missingIDs, reqID)
+		}
+	}
+	return missingIDs
+}
+
+func userGrantMap(record *models.PermissionRecord, nftAddr string, tokenID int64) (map[string]bool, map[string]map[string]*set.StringSet, error) {
+	userPermGrants := make(map[string]bool)
+	cloudEvtGrants := make(map[string]map[string]*set.StringSet)
+
+	if err := validAssetDID(record.Data.Asset, nftAddr, tokenID); err != nil {
+		return nil, nil, fmt.Errorf("failed to validate permission asset: %s", record.Data.Asset)
+	}
+
+	// Aggregates all the permission and attestation grants the user has.
+	for _, agreement := range record.Data.Agreements {
+		now := time.Now()
+		if !agreement.EffectiveAt.IsZero() && now.Before(agreement.EffectiveAt) {
+			continue
+		}
+
+		if !agreement.ExpiresAt.IsZero() && now.After(agreement.ExpiresAt) {
+			continue
+		}
+
+		switch agreement.Type {
+		case "cloudevent":
+			if _, ok := cloudEvtGrants[agreement.EventType]; !ok {
+				cloudEvtGrants[agreement.EventType] = map[string]*set.StringSet{}
+			}
+
+			if _, ok := cloudEvtGrants[agreement.EventType][agreement.Source]; !ok {
+				cloudEvtGrants[agreement.EventType][agreement.Source] = set.NewStringSet()
+			}
+
+			for _, id := range agreement.IDs {
+				cloudEvtGrants[agreement.EventType][agreement.Source].Add(id)
+			}
+
+		case "permission":
+			// Add permissions from this agreement
+			for _, permission := range agreement.Permissions {
+				userPermGrants[permission.Name] = true
+			}
+		}
+	}
+
+	return userPermGrants, cloudEvtGrants, nil
+}
+
+// validAssetDID verifies that the provided DID matches the NFT contract address
+// and token ID specified in the permission token request.
+//
+// Parameters:
+//   - did: The decentralized identifier string to validate, typically in the format "did:nft:..."
+//   - nftContractAddr: NFT contract address from token request and token ID to match against
+//   - tokenID: Token ID from token request
+//
+// Returns:
+//   - bool: true if the DID is valid and matches the request parameters, false otherwise
+//   - error: An error describing why validation failed, or nil if validation succeeded
+func validAssetDID(did string, nftContractAddr string, tokenID int64) error {
+	decodedDID, err := cloudevent.DecodeERC721DID(did)
+	if err != nil {
+		return fmt.Errorf("failed to decode DID: %w", err)
+	}
+
+	if decodedDID.ContractAddress != common.HexToAddress(nftContractAddr) {
+		return fmt.Errorf("DID contract address %s does not match request contract address %s",
+			decodedDID.ContractAddress.Hex(), nftContractAddr)
+	}
+
+	if int64(decodedDID.TokenID.Int64()) != tokenID {
```

Comment:
I'm going to flip this around to
```suggestion
	if decodedDID.TokenID.Cmp(big.NewInt(tokenID)) != 0 {
```
if that's okay.

Reply:
Side note: I don't think we need to cast here, actually.